lib/async_hooks.js

```diff
@@ -211,13 +212,44 @@ class AsyncResource {
 }
 
 const storageList = [];
+const seenLayer = [];
+let depth = 0;
+
+function patchPromiseBarrier(currentResource) {
+  PromiseResolve({
+    then(resolve) {
+      const resource = executionAsyncResource();
+      propagateToStorageLists(resource, currentResource);
+      resolve();
+    }
+  });
+}
+
+function propagateToStorageLists(resource, currentResource) {
+  for (let i = 0; i < storageList.length; ++i) {
+    storageList[i]._propagate(resource, currentResource);
+  }
+}
+
 const storageHook = createHook({
   init(asyncId, type, triggerAsyncId, resource) {
     const currentResource = executionAsyncResource();
     // Value of currentResource is always a non null object
-    for (let i = 0; i < storageList.length; ++i) {
-      storageList[i]._propagate(resource, currentResource);
+    propagateToStorageLists(resource, currentResource);
+
+    if (type === 'PROMISE' && !seenLayer[depth]) {
+      seenLayer[depth] = true;
+      patchPromiseBarrier(currentResource);
     }
+  },
+
+  before(asyncId) {
+    depth++;
+    seenLayer[depth] = false;
+  },
+
+  after(asyncId) {
+    depth--;
   }
```

Comment:
I’m -1 to this fix. This is going to slow everything down because of the use use of before and after hooks.

Reply:
I'm thinking a config option like `trackAsyncAwait` which would switch between two hook sets so it only does the before/after if a user of `AsyncLocalStorage` has explicitly requested it. What do you think?

Reply:
I'm not sure. I think we should investigate if there is another possible fix first, this API is extremely nice and losing so much perf would not be good for the ecosystem.

Reply:
Yep, I'm trying to figure out a lower-level fix. I just made this as a possible higher-level solution for now, until we can come up with something better. Agreed it's not great though, needing the extra before/after hooks.

Reply:
Considering that this problem is not new of AL, I don't think this should be rushed in.

Reply:
Yep, that's fine. Just putting something up that I can iterate on. If the possible lower-level fix is what it needs to be to land, so be it. :)